cmd/tctest/cli/tc.go

```diff
@@ -53,71 +63,153 @@ func TcBuild(server, buildTypeId, branch, testRegEx, user, pass string) (string,
 	}
 
-	req.SetBasicAuth(user, pass)
-	req.Header.Set("Content-Type", "application/xml")
-
-	resp, err := common.Http.Do(req)
-	if err != nil {
-		return "", "", fmt.Errorf("build request failed: %v", err)
-	}
-	defer resp.Body.Close()
-
-	if resp.StatusCode != http.StatusOK {
-		return "", "", fmt.Errorf("HTTP status NOT OK: %d", resp.StatusCode)
+	if statusCode != http.StatusOK {
+		return "", "", fmt.Errorf("HTTP status NOT OK: %d", statusCode)
 	}
 
 	data := struct {
 		BuildId string `xml:"id,attr"`
 	}{}
-	if err := xml.NewDecoder(resp.Body).Decode(&data); err != nil {
-		return "", "", fmt.Errorf("unable to decode XML: %d", resp.StatusCode)
+	if err := xml.NewDecoder(strings.NewReader(body)).Decode(&data); err != nil {
+		return "", "", fmt.Errorf("unable to decode XML: %d", statusCode)
 	}
 
 	return data.BuildId, fmt.Sprintf("https://%s/viewQueued.html?itemId=%s", server, data.BuildId), nil
 }
 
-func TcTestStatus(server, buildId, user, pass string) error {
-	url := fmt.Sprintf("https://%s/downloadBuildLog.html?buildId=%s", server, buildId)
+func TcTestResults(server, buildId, user, pass string, wait bool) error {
+	url := fmt.Sprintf("https://%s/app/rest/builds/%s/state", server, buildId)
+	statusCode, buildStatus, err := makeTcApiCall(url, "", "GET", user, pass)
+	if err != nil {
+		return fmt.Errorf("error looking for build %s state: %v", buildId, err)
+	}
+	if statusCode == http.StatusNotFound {
+		return fmt.Errorf("no build ID %s found in running builds or queue", buildId)
+	}
+	if statusCode != http.StatusOK {
+		return fmt.Errorf("HTTP status NOT OK: %d", statusCode)
+	}
+
+	if buildStatus != "finished" && wait {
+		err := waitForBuild(server, buildId, user, pass)
+		if err != nil {
+			return fmt.Errorf("error waiting for build %s to finish: %v", buildId, err)
+		}
+	}
+	
+	url = fmt.Sprintf("https://%s/downloadBuildLog.html?buildId=%s", server, buildId)
+	statusCode, body, err := makeTcApiCall(url, "", "GET", user, pass)
+	if err != nil {
+		return fmt.Errorf("error looking for build %s results: %v", buildId, err)
+	}
+
+	if statusCode == http.StatusNotFound {
+		// Possibly a queued build, check for it
+		url := fmt.Sprintf("https://%s/app/rest/buildQueue/id:%s", server, buildId)
+		statusCode, _, err := makeTcApiCall(url, "", "GET", user, pass)
+
+		if err != nil {
+			return fmt.Errorf("error checking for build %s in queue: %v", buildId, err)
+		}
+
+		if statusCode == http.StatusNotFound {
+			return fmt.Errorf("no build ID %s found in running builds or queue", buildId)
+		}
+		if statusCode != http.StatusOK {
+			return fmt.Errorf("HTTP status NOT OK: %d", statusCode)
+		}
+		return fmt.Errorf("build %s still queued, check results later...", buildId)
+	}
+	if statusCode != http.StatusOK {
+		return fmt.Errorf("HTTP status NOT OK: %d", statusCode)
+	}
 
-	req, err := http.NewRequest("GET", url, nil)
+	r := regexp.MustCompile(`^\s*--- (FAIL|PASS|SKIP):`)
+	for _, line := range strings.Split(body, "\n") {
```

Comment:
How hard would it be to output the reason for test failures here?

Reply:
I think reasonably difficult, unfortunately - because of how the build log is structured (or not, as the case may be), the reasons for failures will be "some number of lines of test output/stack trace/etc" that would not be trivial to parse out. Not impossible, but annoying enough that it would feel out of scope for this PR.